**Deep Leaning**

In [0]:
from google.colab import drive
drive.mount('/content/drive')

In [2]:
cd /content/drive/My Drive/Datasets/Others

/content/drive/My Drive/Datasets/Others


In [3]:
ls

Churn_Modelling.csv


# ANN

# Preprocessing

In [0]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [0]:
dataset = pd.read_csv("Churn_Modelling.csv")

In [6]:
dataset.head()

RowNumber  CustomerId   Surname  ...  IsActiveMember EstimatedSalary Exited
0          1    15634602  Hargrave  ...               1       101348.88      1
1          2    15647311      Hill  ...               1       112542.58      0
2          3    15619304      Onio  ...               0       113931.57      1
3          4    15701354      Boni  ...               0        93826.63      0
4          5    15737888  Mitchell  ...               1        79084.10      0

[5 rows x 14 columns]

In [7]:
dataset.describe

<bound method NDFrame.describe of       RowNumber  CustomerId    Surname  ...  IsActiveMember EstimatedSalary Exited
0             1    15634602   Hargrave  ...               1       101348.88      1
1             2    15647311       Hill  ...               1       112542.58      0
2             3    15619304       Onio  ...               0       113931.57      1
3             4    15701354       Boni  ...               0        93826.63      0
4             5    15737888   Mitchell  ...               1        79084.10      0
...         ...         ...        ...  ...             ...             ...    ...
9995       9996    15606229   Obijiaku  ...               0        96270.64      0
9996       9997    15569892  Johnstone  ...               1       101699.77      0
9997       9998    15584532        Liu  ...               1        42085.58      1
9998       9999    15682355  Sabbatini  ...               0        92888.52      1
9999      10000    15628319     Walker  ...          

In [0]:
X = dataset.iloc[:,3:13].values
y = dataset.iloc[:,13].values

In [9]:
X

array([[619, 'France', 'Female', ..., 1, 1, 101348.88],
       [608, 'Spain', 'Female', ..., 0, 1, 112542.58],
       [502, 'France', 'Female', ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 'Female', ..., 0, 1, 42085.58],
       [772, 'Germany', 'Male', ..., 1, 0, 92888.52],
       [792, 'France', 'Female', ..., 1, 0, 38190.78]], dtype=object)

In [0]:
#for converting from categorical to numerical like word to number4
from sklearn.preprocessing import LabelEncoder,OneHotEncoder
from sklearn.compose import ColumnTransformer

In [0]:
# for gender
labelEncoderGender = LabelEncoder()
X[:,2] = labelEncoderGender.fit_transform(X[:,2])

In [12]:
X

array([[619, 'France', 0, ..., 1, 1, 101348.88],
       [608, 'Spain', 0, ..., 0, 1, 112542.58],
       [502, 'France', 0, ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 0, ..., 0, 1, 42085.58],
       [772, 'Germany', 1, ..., 1, 0, 92888.52],
       [792, 'France', 0, ..., 1, 0, 38190.78]], dtype=object)

In [0]:
transformer = ColumnTransformer(
    transformers=[
        ("OneHot",        # Just a name
         OneHotEncoder(), # The transformer class
         [1]              # The column(s) to be applied on.
         )
    ],
    remainder='passthrough' # donot apply anything to the remaining columns
)
X = transformer.fit_transform(X)

In [14]:
X

array([[1.0, 0.0, 0.0, ..., 1, 1, 101348.88],
       [0.0, 0.0, 1.0, ..., 0, 1, 112542.58],
       [1.0, 0.0, 0.0, ..., 1, 0, 113931.57],
       ...,
       [1.0, 0.0, 0.0, ..., 0, 1, 42085.58],
       [0.0, 1.0, 0.0, ..., 1, 0, 92888.52],
       [1.0, 0.0, 0.0, ..., 1, 0, 38190.78]], dtype=object)

In [0]:
X = X[:,1:]

In [16]:
X

array([[0.0, 0.0, 619, ..., 1, 1, 101348.88],
       [0.0, 1.0, 608, ..., 0, 1, 112542.58],
       [0.0, 0.0, 502, ..., 1, 0, 113931.57],
       ...,
       [0.0, 0.0, 709, ..., 0, 1, 42085.58],
       [1.0, 0.0, 772, ..., 1, 0, 92888.52],
       [0.0, 0.0, 792, ..., 1, 0, 38190.78]], dtype=object)

In [0]:
X[:,0] = X[:,0].astype(int)
X[:,1] = X[:,1].astype(int)

In [18]:
X

array([[0, 0, 619, ..., 1, 1, 101348.88],
       [0, 1, 608, ..., 0, 1, 112542.58],
       [0, 0, 502, ..., 1, 0, 113931.57],
       ...,
       [0, 0, 709, ..., 0, 1, 42085.58],
       [1, 0, 772, ..., 1, 0, 92888.52],
       [0, 0, 792, ..., 1, 0, 38190.78]], dtype=object)

In [0]:
#training start (spliting)
from sklearn.model_selection import train_test_split

In [0]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [0]:
# scaling features
from sklearn.preprocessing import StandardScaler


In [0]:
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

# Training and tesing

In [0]:
import keras
from keras.models import Sequential
from keras.layers import Dense

In [0]:
classifier = Sequential()

In [0]:
#dense will initialize the input and hidden layer in ann
#selecting output dim is an Art , no rules of thumb, for this session - I am taking average of input and output layer  which is 11+1 / 2 = 6
#init initialized the initial weight, which we need to initialize close to Zero  with normal / uniform distribution
#in standard way, relu is good for input and sigmoid is good for output   
#input dimension is first attempt must have to be mention 
#important note is , in dense we  must declare all the necessary parameter
classifier.add(Dense(output_dim = 6, init= 'uniform', activation = 'relu', input_dim = 11))

In [0]:
#as first input and hidden layer is initialized with Dense()
#second hidden layer
# as I've mentioned earlier , no need  to add input_dim after first attempt , as  the system already knows about this 
classifier.add(Dense(output_dim = 6, init= 'uniform', activation = 'relu'))

In [0]:
#output layer
#sigmoid is  the heart of probabilistic function
#it is  a binary classification so output dim here will be = 1
#for more than binary classification , such as for 3 categories we'll need  3 output dim
#and activation will be "softmax"
classifier.add(Dense(output_dim = 1, init= 'uniform', activation = 'sigmoid'))

In [0]:
#optimization stuffs
#for backpropagations stuffs  I have used "adam" here, loss function => for binary = binary_crossentropy ->for more than two = categorical_crossentropy  (sgd)
#after each update of the weight we need  some means to evaluate the model at present situation
#so for that method funcstion come metrics= ['accuracy'] (as it demands a  list)
classifier.compile(optimizer='adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

In [39]:
  #here we have only defined  layers but till now  no connection -> for  this fit actually comes in
  #for selecting batch_size and  epochs , there's no fixed rules 
  classifier.fit(X_train,y_train, batch_size = 10, epochs= 100)




Epoch 1/100





8000/8000 [==============================] - 12s 2ms/step - loss: 0.5039 - acc: 0.7957
Epoch 2/100
8000/8000 [==============================] - 3s 326us/step - loss: 0.4301 - acc: 0.7960
Epoch 3/100
8000/8000 [==============================] - 3s 331us/step - loss: 0.4251 - acc: 0.7960
Epoch 4/100
8000/8000 [==============================] - 3s 331us/step - loss: 0.4215 - acc: 0.7960
Epoch 5/100
8000/8000 [==============================] - 3s 334us/step - loss: 0.4186 - acc: 0.8141
Epoch 6/100
8000/8000 [==============================] - 3s 326us/step - loss: 0.4162 - acc: 0.8256
Epoch 7/100
8000/8000 [==============================] - 3s 333us/step - loss: 0.4145 - acc: 0.8282
Epoch 8/100
8000/8000 [==============================] - 3s 334us/step - loss: 0.4132 - acc: 0.8312
Epoch 9/100
8000/8000 [==============================] - 3s 327us/step - loss: 0.4117 - acc: 0.8324
Epoch 10/100
8000/8000 [==============================] - 3s 334us/step - loss: 0.4109 - acc:

In [0]:
y_pred = classifier.predict(X_test)

In [0]:
#y_pred contains the probabilities , so for determining the  final decision
threshold = 0.5
y_pred = y_pred > threshold

In [0]:
from sklearn.metrics import confusion_matrix
confMat = confusion_matrix(y_test,y_pred)

In [61]:
confMat

array([[1549,   46],
       [ 275,  130]])

In [62]:
(1549+130) / 2000

0.8395

In [70]:
#single evaluation point
new_evaluationPoint = np.array([[0,0,600,1,40,3,60000,2,1,1,50000]])
scaledPoint = sc.transform(new_evaluationPoint)
new_pred = classifier.predict(scaledPoint)
new_pred = new_pred > 0.5
new_pred

array([[False]])

So with ANN we have accuracy of 84% (approx) without using any tuning
And as training and testing accuracy is almost similar , so this is satisfactory 

# CNN